hibernate-core/src/main/java/org/hibernate/engine/internal/StatefulPersistenceContext.java

```diff
@@ -423,56 +423,54 @@ public boolean containsEntityHolder(EntityKey key) {
 
 	@Override
 	public void postLoad(JdbcValuesSourceProcessingState processingState, Consumer<EntityHolder> holderConsumer) {
-		if ( entitiesByKey == null ) {
+		if ( CollectionHelper.isEmpty( processingState.getLoadingEntityHolders() ) ) {
 			return;
 		}
 		final Callback callback = processingState.getExecutionContext().getCallback();
 		final EventListenerGroup<PostLoadEventListener> listenerGroup = getSession().getFactory()
 				.getFastSessionServices()
 				.eventListenerGroup_POST_LOAD;
-		final PostLoadEvent postLoadEvent = getSession().isEventSource() ? new PostLoadEvent( getSession().asEventSource() ) : null;
-		for ( Iterator<EntityHolderImpl> iterator = entitiesByKey.values().iterator(); iterator.hasNext(); ) {
-			final EntityHolderImpl holder = iterator.next();
-			if ( holder.processingState == processingState ) {
-				if ( holderConsumer != null ) {
-					holderConsumer.accept( holder );
-				}
-				if ( holder.entity == null ) {
-					// It's possible that we tried to load an entity and found out it doesn't exist,
-					// in which case we added an entry with a null proxy and entity.
-					// Remove that empty entry on post load to avoid unwanted side effects
-					iterator.remove();
-					continue;
-				}
-				if ( postLoadEvent != null ) {
-					postLoadEvent.reset();
-					postLoadEvent.setEntity( holder.entity )
-							.setId( holder.entityKey.getIdentifier() )
-							.setPersister( holder.getDescriptor() );
-					listenerGroup.fireEventOnEachListener(
-							postLoadEvent,
-							PostLoadEventListener::onPostLoad
-					);
-				}
+		final PostLoadEvent postLoadEvent = processingState.getPostLoadEvent();
+		final Iterator<EntityHolder> iterator = processingState.getLoadingEntityHolders().iterator();
+		while ( iterator.hasNext() ) {
+			final EntityHolder holder = iterator.next();
```

Comment:
Let's simplify this a bit further. No need for an iterator since the processingState is going to be garbage collected after this method anyway.

```suggestion
		for ( EntityHolder holder : processingState.getLoadingEntityHolders() ) {
```

Reply:
That's actually what I did initially, but there were failing tests. Apprently, in e.g. `ScrollableResultsImpl#prepareCurrentRow`, `StatefulPersistenceContext#postLoad` is called after reading every row from the database, so if there are multiple results `getLoadingEntityHolders()` would contain every previous row. I had to use an iterator and call `iterator.remove()` at the end to avoid processing the same result multiple times.

Reply:
Updated with `.clear()` to efficiently empty the collection and a separate list for reloaded entity holders.